### PR TITLE
Add generic parameter to AwilixContainer interface

### PR DESCRIPTION
**Pull request by roikoren755**

### CODE REVIEW

File: src/container.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ export interface AwilixContainer {
   /**
    * Creates a scoped container with this one as the parent.
    */
-  createScope(): AwilixContainer
+  createScope<T extends object = any>(): AwilixContainer<T>
```
Comment (jeffijoe) [inline src/container.ts]: Shouldn't this be

```suggestion
  createScope<T extends object = any>(): AwilixContainer<Cradle & T>
```

File: src/container.ts [hunk 2]
```diff
@@ -122,24 +122,26 @@ export interface ResolveOptions {
...
 /**
  * Register a Registration
  * @interface NameAndRegistrationPair
  */
-export interface NameAndRegistrationPair {
-  [key: string]: Resolver<any>
-}
+export type NameAndRegistrationPair<T> = Partial<
+  {
+    [U in keyof T]: Resolver<T[U]>
```
Comment (jeffijoe) [inline src/container.ts]: Remove `Partial` and use

```suggestion
    [U in keyof T]?: Resolver<T[U]>
```

File: src/load-modules.ts [hunk 3]
```diff
@@ -43,9 +43,9 @@ export type NameFormatter = (
 /**
  * Dependencies for `loadModules`
  */
-export interface LoadModulesDeps {
+export interface LoadModulesDeps<T extends object> {
```
Comment (jeffijoe) [inline src/load-modules.ts]: `loadModules` is dynamic already, I don't think we need to worry about it here.

File: src/container.ts (hunk 1 above)
Comment (roikoren755) [reply src/container.ts]: Yep, my bad

File: src/container.ts (hunk 2 above)
Comment (roikoren755) [reply src/container.ts]: 👍

File: src/load-modules.ts (hunk 3 above)
Comment (roikoren755) [reply src/load-modules.ts]: Don't need to worry in this file altogether? or just this interface and its uses?

Comment (jeffijoe) [reply src/load-modules.ts]: I think the entire file. Loading modules from the file system can never provide any static type analysis.

**Comment by jeffijoe** [comment]
One last favor before we release, could you check out `awilix-koa` and link your local `awilix` and run the lint + test? Want to confirm that this isn't a breaking change type-wise, as I don't want to cut a major release just for a type change. 🙏

File: src/load-modules.ts (hunk 3 above)
Comment (roikoren755) [reply src/load-modules.ts]: Reverted this file

**Comment by roikoren755** [comment]
> One last favor before we release, could you check out `awilix-koa` and link your local `awilix` and run the lint + test? Want to confirm that this isn't a breaking change type-wise, as I don't want to cut a major release just for a type change. 🙏

Just did, and everything passed. Just to be sure, what I did to test it was change the `awilix` entries in the `awilix-koa` `package.json` to be `file:../awilix`, and then ran `yarn lint` and `yarn test`

**Comment by jeffijoe** [comment]
Sounds good, last step is squashing your commits into 1 commit. If you don't want to, I can do a Squash and Merge on GH but it's not as nice. 😄

**Comment by roikoren755** [comment]
Done :)

**Comment by jeffijoe** [comment]
Fantastic, thank you!

**Pull request merged**